compiler/src/org.graalvm.compiler.core.test/src/org/graalvm/compiler/core/test/OptimizedBoxNodeTest.java

```diff
@@ -208,14 +216,92 @@ public void testType(String typePrefix, int lowBound, int highBound, BoxProducer
             if (originalVal >= cacheLow && originalVal <= cacheHigh) {
                 // in cache, all must be the same objects
                 Assert.assertTrue("val=" + originalVal + " optimized version must remain cached object identities", interpreterObject == objectReuse);
-                Assert.assertTrue("val=" + originalVal + " unoptimized version must remain cached object identities", interpreterObject == objectNoReuse);
             } else {
-                Assert.assertTrue("val=" + originalVal + " out of cache, unoptimized version must not reuse the argument from the call and thus be different than the interpreter object",
-                                interpreterObject != objectNoReuse);
                 Assert.assertTrue("val=" + originalVal + " out of cache, optimized version must re-use the argument from the call and thus be the same as the interpreter object",
                                 interpreterObject == objectReuse);
             }
         }
+    }
+
+    static int snippetConstantCompare(int a) {
+        int res = a;
+        for (int i = -200; i < 200; i++) {
+            if (a(i) == b(i + 1)) {
+                GraalDirectives.sideEffect(1);
+                res = i;
+            }
+            GraalDirectives.sideEffect(123);
+        }
+        GraalDirectives.sideEffect(2);
+        return res;
+    }
+
+    static Integer a(int i) {
+        GraalDirectives.sideEffect(3);
+        return Integer.valueOf(i);
+    }
+
+    static Integer b(int i) {
+        GraalDirectives.sideEffect(4);
+        return Integer.valueOf(i);
+    }
+
+    @Test
+    public void testCompare() throws InvalidInstalledCodeException {
+        final OptionValues testOptions = new OptionValues(getInitialOptions(), DefaultLoopPolicies.Options.FullUnrollMaxNodes, 10000, DefaultLoopPolicies.Options.ExactFullUnrollMaxNodes, 10000);
+        StructuredGraph g = parseEager(getResolvedJavaMethod("snippetConstantCompare"), AllowAssumptions.NO, testOptions);
+        CanonicalizerPhase.create().apply(g, getDefaultHighTierContext());
+        new LoopFullUnrollPhase(createCanonicalizerPhase(), new DefaultLoopPolicies()).apply(g, getDefaultHighTierContext());
+        CanonicalizerPhase.create().apply(g, getDefaultHighTierContext());
+        Assert.assertEquals("All ifs must be removed", 0, g.getNodes(IfNode.TYPE).count());
+        int[] res = new int[200 * 2];
+        int[] resCompiled = new int[200 * 2];
+        for (int i = -200; i < 200; i++) {
+            res[i + 200] = snippetConstantCompare(i);
+        }
+        InstalledCode code = getCode(getResolvedJavaMethod("snippetConstantCompare"), testOptions);
+        for (int i = -200; i < 200; i++) {
+            resCompiled[i + 200] = (int) code.executeVarargs(i);
+        }
+        Assert.assertArrayEquals(res, resCompiled);
+    }
+
+    static int testPEASnippet() {
+        if (Integer.valueOf(1024) == Integer.valueOf(1024)) {
```

Comment:
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [SA: Self comparison of value with itself (SA_LOCAL_SELF_COMPARISON)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#sa-self-comparison-of-value-with-itself-sa-local-self-comparison).
This method compares a local variable with itself, and may indicate a typo or a logic error. Make sure that you are comparing the right things.